src/swapRouter.ts

```diff
@@ -175,62 +180,203 @@ export abstract class SwapRouter {
+      const amountIn: string = toHex(trade.maximumAmountIn(options.slippageTolerance, inputAmount).quotient)
+      const amountOut: string = toHex(trade.minimumAmountOut(options.slippageTolerance, outputAmount).quotient)
+
+      // flag for whether the trade is single hop or not
+      const singleHop = route.pools.length === 1
+
+      const recipient = routerMustCustody
+        ? ADDRESS_THIS
+        : typeof options.recipient === 'undefined'
+        ? MSG_SENDER
+        : validateAndParseAddress(options.recipient)
+
+      const mixedRouteIsAllV3 = (route: MixedRoute<Currency, Currency>) => {
+        return route.pools.every((pool) => pool instanceof Pool)
+      }
+
+      if (singleHop) {
+        /// For single hop, since it isn't really a mixedRoute, we'll just mimic behavior of V3 or V2
+        /// We don't use encodeV3Swap or encodeV2Swap because casting the trade to a V3Trade or V2Trade is overcomplicated
+        if (route.pools.every((pool) => pool instanceof Pool)) {
+          const exactInputSingleParams = {
+            tokenIn: route.path[0].address,
+            tokenOut: route.path[1].address,
+            fee: (route.pools as Pool[])[0].fee,
+            recipient,
+            amountIn,
+            amountOutMinimum: performAggregatedSlippageCheck ? 0 : amountOut,
+            sqrtPriceLimitX96: 0,
+          }
+
+          calldatas.push(SwapRouter.INTERFACE.encodeFunctionData('exactInputSingle', [exactInputSingleParams]))
+        } else {
+          const path = trade.route.path.map((token) => token.address)
+
+          const exactInputParams = [amountIn, performAggregatedSlippageCheck ? 0 : amountOut, path, recipient]
+
+          calldatas.push(SwapRouter.INTERFACE.encodeFunctionData('swapExactTokensForTokens', exactInputParams))
+        }
+      } else {
+        const sections = divideMixedRouteIntoConsecutiveSections(route)
+
+        const firstInputToken = route.input.wrapped
+        let outputToken = firstInputToken
+
+        for (let i = 0; i < sections.length; i++) {
+          const isLastSectionInRoute = (i: number) => {
+            return i === sections.length - 1
+          }
+
+          const section = sections[i]
+
+          /// Previous output is now input, save this before reassigning
+          let nextInput = outputToken
+          /// Now, we get output of this section
+          outputToken = getOutputOfPools(section, outputToken)
+
+          const newRouteOriginal = new MixedRouteSDK(
+            [...section],
+            section[0].token0.equals(nextInput) ? section[0].token0 : section[0].token1,
+            outputToken
+          )
+          const newRoute = new MixedRoute(newRouteOriginal)
+          const path: string = encodeMixedRouteToPath(newRoute)
+
+          if (mixedRouteIsAllV3(newRoute)) {
+            const exactInputParams = {
+              path,
+              /// routerMustCustody is prioritized
```

Comment:
Can we make this comment more clear like `router holds funds until the last swap` or something like that. i also think the docs at the params will help

Reply:
After chatting offline, maybe add this to the note above since that is referring to a specific case

Reply:
```
              // By default router holds funds until the last swap, then it is sent to the recipient
              // special case exists where we are unwrapping WETH output, in which case `routerMustCustody` is set to true
              // and router still holds the funds. That logic bundled into how the value of `recipient` is calculated
```